play-json/shared/src/main/scala/Reads.scala

```diff
@@ -373,34 +379,89 @@ trait DefaultReads extends LowPriorityDefaultReads {
   /**
    * Deserializer for Map[String,V] types.
    */
-  implicit def mapReads[V](implicit fmtv: Reads[V]): Reads[collection.immutable.Map[String, V]] = new Reads[collection.immutable.Map[String, V]] {
-    def reads(json: JsValue) = json match {
-      case JsObject(m) => {
+  implicit def mapReads[V](implicit fmtv: Reads[V]): Reads[collection.immutable.Map[String, V]] = Reads[collection.immutable.Map[String, V]] {
+    case JsObject(m) => {
+      type Errors = Seq[(JsPath, Seq[JsonValidationError])]
+      def locate(e: Errors, key: String) = e.map {
+        case (p, valerr) => (JsPath \ key) ++ p -> valerr
+      }
 
-        type Errors = Seq[(JsPath, Seq[JsonValidationError])]
-        def locate(e: Errors, key: String) = e.map {
-          case (p, valerr) => (JsPath \ key) ++ p -> valerr
+      m.foldLeft(Right(Map.empty): Either[Errors, Map[String, V]]) {
+        case (acc, (key, value)) => (acc, fmtv.reads(value)) match {
+          case (Right(vs), JsSuccess(v, _)) => Right(vs + (key -> v))
+          case (Right(_), JsError(e)) => Left(locate(e, key))
+          case (Left(e), _: JsSuccess[_]) => Left(e)
+          case (Left(e1), JsError(e2)) => Left(e1 ++ locate(e2, key))
         }
-
-        m.foldLeft(Right(Map.empty): Either[Errors, Map[String, V]]) {
-          case (acc, (key, value)) => (acc, fmtv.reads(value)) match {
-            case (Right(vs), JsSuccess(v, _)) => Right(vs + (key -> v))
-            case (Right(_), JsError(e)) => Left(locate(e, key))
-            case (Left(e), _: JsSuccess[_]) => Left(e)
-            case (Left(e1), JsError(e2)) => Left(e1 ++ locate(e2, key))
-          }
-        }.fold(JsError.apply, res => JsSuccess(res.toMap))
-      }
-      case _ => JsError(Seq(JsPath -> Seq(JsonValidationError("error.expected.jsobject"))))
+      }.fold(JsError.apply, res => JsSuccess(res.toMap))
     }
+
+    case _ => JsError(Seq(JsPath -> Seq(
+      JsonValidationError("error.expected.jsobject"))))
   }
 
   /**
    * Deserializer for Array[T] types.
    */
-  implicit def ArrayReads[T: Reads: ClassTag]: Reads[Array[T]] = new Reads[Array[T]] {
-    def reads(json: JsValue) = json.validate[List[T]].map(_.toArray)
-  }
+  implicit def ArrayReads[T: Reads: ClassTag]: Reads[Array[T]] =
+    Reads[Array[T]] { _.validate[List[T]].map(_.toArray) }
+
+  private def finiteDurationNumberReads(esuffix: String) =
+    Reads[FiniteDuration] {
+      case JsNumber(n) => JsSuccess {
+        val millis = if (!n.ulp.isWhole) {
+          (n.toDouble * 1000).toLong
+        } else n.toLong
```

Comment:
I'm also ok with doing some sort of rounding/truncation, but it should be consistent for all numbers.